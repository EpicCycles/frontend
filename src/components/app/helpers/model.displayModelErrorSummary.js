import {customerAddressFields} from "../model/helpers/fields";
import {displayModelErrorSummary} from "../model/helpers/model";

test('if no error or error_detail undefined returned', () => {
    const model = {
        id: 123,
        address1: "line one",
        address2: "line Ywo",
        address3: "line Three",
        address4: "line Four",
        postcode: "xxxyyy",
        customer: 6
    };
    const result = displayModelErrorSummary(model, customerAddressFields);
    expect(result).toBe(undefined);
});
test('if just error no error_detail error returned', () => {
    const model = {
        id: 123,
        address1: "line one",
        address2: "line Ywo",
        address3: "line Three",
        address4: "line Four",
        postcode: "xxxyyy",
        customer: 6,
        error: "Model level error"
    };
    const result = displayModelErrorSummary(model, customerAddressFields);
    expect(result).toBe("Model level error");
});
test('if no error with error_detail error_details formatted returned', () => {
    const model = {
        id: 123,
        address1: "line one",
        address2: "line Ywo",
        address3: "line Three",
        address4: "line Four",
        postcode: "xxxyyy",
        customer: 6,
        error_detail: {
            "address1": ["A valid address is required.", "SHow me the address is required."],
            "postcode": ["This field may not be blank."]
        },
    };
    const expectedText = "Address: A valid address is required. SHow me the address is required.<br>Postcode: This field may not be blank.";
    const result = displayModelErrorSummary(model, customerAddressFields);
    expect(result).toBe(expectedText);
});
test('if error and error_detail formatted returned', () => {
    const model = {
        id: 123,
        address1: "line one",
        address2: "line Ywo",
        address3: "line Three",
        address4: "line Four",
        postcode: "xxxyyy",
        customer: 6,
        error: "Model level error",
        error_detail: {
            "address1": ["A valid address is required.", "SHow me the address is required."],
            "postcode": ["This field may not be blank."]
        },
    };
    const expectedText = "Model level error<br>Address: A valid address is required. SHow me the address is required.<br>Postcode: This field may not be blank.";
    const result = displayModelErrorSummary(model, customerAddressFields);
    expect(result).toBe(expectedText);
});