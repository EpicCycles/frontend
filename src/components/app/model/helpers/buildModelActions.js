/**
 * buildModelActions - builds actions for the current model (use to build additional actions)
 * @param requiredActions - array of objects {iconName, iconTitle, requiresIdOnly, actionFunction}
 * @param model - model being displayed oredited
 * @returns array of model actions specific to current model
 */
export const buildModelActions = (requiredActions, model) => {
  return requiredActions.map(action => {
    const finalAction = {
      iconName: action.iconName,
      iconTitle: action.iconTitle,
    };
    const isSavedModel = model && model.id;
    if (action.requiresIdOnly) {
      finalAction.iconAction = () => isSavedModel && action.actionFunction(model.id);
    } else {
      finalAction.iconAction = () => isSavedModel && action.actionFunction(model);
    }
    finalAction.iconDisabled = !isSavedModel;
    return finalAction;
  });
};
