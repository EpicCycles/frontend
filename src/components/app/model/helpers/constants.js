export const POSTCODE_RULES = [
  { countryCode: 'AT', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'AU', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'BE', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'BG', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'CA', regex: /^[A-Za-z][0-9][A-Za-z] [0-9][A-Za-z][0-9]/, display: 'ANA NAN' },
  { countryCode: 'CH', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'CY', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'CZ', regex: /^\d{3}\s\d{2}$/, display: 'NNN NN' },
  { countryCode: 'DE', regex: /^\d{5}$/, display: 'NNNNN' },
  { countryCode: 'DK', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'EE', regex: /^\d{5}$/, display: 'NNNNN' },
  { countryCode: 'ES', regex: /^\d{5}$/, display: 'NNNNN' },
  { countryCode: 'FI', regex: /^\d{5}$/, display: 'NNNNN' },
  { countryCode: 'FR', regex: /^\d{5}$/, display: 'NNNNN' },
  {
    countryCode: 'GB',
    regex: /^[A-Z](([0-9][0-9]?)|([A-Z][0-9][0-9]?)|([A-Z]?[0-9][A-Z])) ?[0-9][ABD-HJLNP-UW-Z]{2}$/,
    display: 'CCNN NCC',
  },
  { countryCode: 'GR', regex: /^\d{3}\s\d{2}$/, display: 'NNNNN or NNN NN' },
  { countryCode: 'HU', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'IT', regex: /^\d{5}$/, display: 'NNNNN' },
  { countryCode: 'KR', regex: /^\d{5}$/, display: 'NNNNN' },
  { countryCode: 'LT', regex: /^\d{5}$/, display: 'NNNNN' },
  { countryCode: 'LU', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'MT', regex: /^\[A-Z]{3}\s\d{4}$/, display: 'AAA NNNN' },
  { countryCode: 'MY', regex: /^\d{5}$/, display: 'NNNNN' },
  { countryCode: 'NO', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'NZ', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'PH', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'PL', regex: /^\d{2}-\d{3}$/, display: 'NN-NNN' },
  { countryCode: 'PT', regex: /^\d{4}-\d{3}$/, display: 'NNNN-NNN' },
  { countryCode: 'RO', regex: /^\d{6}$/, display: 'NNNNNN' },
  { countryCode: 'SE', regex: /^\d{3,4}$/, display: 'NNNNN or NNNN' },
  { countryCode: 'SG', regex: /^\d{6}$/, display: 'NNNNNN' },
  { countryCode: 'SI', regex: /^\d{3}\s\d{2}$/, display: 'NNN NN' },
  { countryCode: 'SK', regex: /^\d{3}\s\d{2}$/, display: 'NNN NN' },
  { countryCode: 'US', regex: /^(\d{5}([-]\d{4})?)$/, display: 'NNNNN-NNNN or NNNNN' },
  { countryCode: 'ZA', regex: /^\d{4}$/, display: 'NNNN' },
];
