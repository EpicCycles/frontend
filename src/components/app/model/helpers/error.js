export const BRAND_NAME_MISSING = 'A name is required for the brand';
export const FRAME_NAME_MISSING = 'A name is required for the frame';
export const MODEL_NAME_MISSING = 'A name is required for the model';
export const PART_NAME_MISSING = 'A name is required for the part';
export const PART_TYPE_NAME_MISSING = 'A name is required for the part type';
export const SECTION_MISSING = 'A section must be selected for the part type';
export const PART_TYPE_MISSING = 'A part type must be selected';
export const BRAND_MISSING = 'A brand must be selected';
export const SUPPLIER_MISSING = 'A supplier must be selected';
export const ADDRESS_MISSING = 'At least one line of address must be entered';
export const COUNTRY_MISSING = 'A country must be selected';
export const INVALID_POSTCODE = 'A postcode is expected which for this country has the format ';
export const SELECT_ONE_MISSING = 'Please select an option.';
export const VALUE_MISSING = 'A value is required.';
export const INVALID_EMAIL = 'Invalid email address.';
export const INVALID_URL = 'Invalid URL';
export const INVALID_NUMBER = 'This should be a number';
export const INVALID_INTEGER = 'This should be a number greater than zero';
export const INVALID_CURRENCY = 'This should be a number with up to two decimal places';
