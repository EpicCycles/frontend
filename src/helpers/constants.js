export const NEW_ELEMENT_ID = 'new';

export const colourStyles = [
  { background: 'bg-col-01', colour: 'col-01', border: 'border-col-01' },
  {
    background: 'bg-col-02',
    colour: 'col-02',
    border: 'border-col-02',
    transition: 'transition-01-02',
  },
  {
    background: 'bg-col-03',
    colour: 'col-03',
    border: 'border-col-03',
    transition: 'transition-02-03',
  },
  {
    background: 'bg-col-04',
    colour: 'col-04',
    border: 'border-col-04',
    transition: 'transition-03-04',
  },
  {
    background: 'bg-col-05',
    colour: 'col-05',
    border: 'border-col-05',
    transition: 'transition-04-05',
  },
  {
    background: 'bg-col-06',
    colour: 'col-06',
    border: 'border-col-06',
    transition: 'transition-05-06',
  },
  {
    background: 'bg-col-07',
    colour: 'col-07',
    border: 'border-col-07',
    transition: 'transition-06-07',
  },
  {
    background: 'bg-col-08',
    colour: 'col-08',
    border: 'border-col-08',
    transition: 'transition-07-08',
  },
  {
    background: 'bg-col-09',
    colour: 'col-09',
    border: 'border-col-09',
    transition: 'transition-08-09',
  },
  {
    background: 'bg-col-10',
    colour: 'col-10',
    border: 'border-col-10',
    transition: 'transition-09-10',
  },
  {
    background: 'bg-col-11',
    colour: 'col-11',
    border: 'border-col-11',
    transition: 'transition-10-11',
  },
  {
    background: 'bg-col-12',
    colour: 'col-12',
    border: 'border-col-12',
    transition: 'transition-11-12',
  },
  {
    background: 'bg-col-13',
    colour: 'col-13',
    border: 'border-col-13',
    transition: 'transition-12-13',
  },
];

export const STORAGE_PARTS = 'epic_parts';
export const STORAGE_SUPPLIER_PRODUCTS = 'epic_supplierProducts';
export const STORAGE_SECTIONS = 'epic_sections';
export const STORAGE_BRANDS = 'epic_brands';
export const STORAGE_CHARGES = 'epic_charges';
export const STORAGE_SUPPLIERS = 'epic_suppliers';
export const COOKIE_USER = 'epic_user';
export const COOKIE_TOKEN = 'epic_user_token';
