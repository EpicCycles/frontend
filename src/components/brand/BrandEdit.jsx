import React from 'react';
import * as PropTypes from 'prop-types';

import FormTextInput from '../../common/FormTextInput';
import { generateRandomCode, removeObjectWithIndex, updateObject } from '../../helpers/utils';
import { Icon } from 'semantic-ui-react';
import { NEW_ELEMENT_ID } from '../../helpers/constants';
import SupplierBlob from '../supplier/SupplierBlob';
import { brandFields } from '../app/model/helpers/fields';
import { updateModel } from '../app/model/helpers/model';

class BrandEdit extends React.Component {
  handleBrandValueChange = (fieldName, input) => {
    const updatedBrand = updateModel(this.props.brand, brandFields, fieldName, input);

    if (this.props.componentKey === NEW_ELEMENT_ID) updatedBrand.dummyKey = NEW_ELEMENT_ID;
    this.props.handleBrandChange(this.props.componentKey, updatedBrand);
  };
  removeSupplier = supplierKey => {
    let currentSuppliers = this.props.brand.supplier || [];
    const supplierIndex = currentSuppliers.indexOf(supplierKey);
    if (supplierIndex > -1) {
      const updatedSupplier = removeObjectWithIndex(currentSuppliers, supplierIndex);

      this.handleBrandValueChange('supplier', updatedSupplier);
    }
  };

  handleInputClear = fieldName => {
    if (window.confirm('Please confirm that you want to delete this Brand')) {
      const updatedBrand = updateObject(this.props.brand);
      updatedBrand.delete = true;
      this.props.handleBrandChange(this.props.componentKey, updatedBrand);
    }
  };

  addAnother = () => {
    const updatedBrand = updateObject(this.props.brand);
    updatedBrand.dummyKey = generateRandomCode();
    this.props.handleBrandChange(NEW_ELEMENT_ID, updatedBrand);
  };

  render() {
    const { brand, componentKey, pickUpBrand, suppliers } = this.props;
    const supplierForBrand = suppliers.filter(
      supplier => brand.supplier && brand.supplier.includes(supplier.id),
    );
    return (
      <div
        key={`brand${componentKey}`}
        className="rounded"
        draggable={pickUpBrand && componentKey !== NEW_ELEMENT_ID}
        onDragStart={event => pickUpBrand(event, componentKey)}
      >
        {componentKey === NEW_ELEMENT_ID && <Icon name="add" onClick={this.addAnother} />}
        <FormTextInput
          placeholder="add new"
          fieldName={`brand_name_${componentKey}`}
          value={brand.brand_name}
          onChange={this.handleBrandValueChange}
          onClick={this.handleInputClear}
        />
        <div className="ui toggle checkbox">
          <input
            type="checkbox"
            name={`bike_brand_${componentKey}`}
            onChange={() => this.handleBrandValueChange('bike_brand', !brand.bike_brand)}
            checked={brand.bike_brand}
          />
          <label>Bike Brand</label>
        </div>
        Supplier(s):{' '}
        {supplierForBrand.length > 0
          ? supplierForBrand.map(supplier => (
              <SupplierBlob
                key={`supplier${componentKey}${supplier.id}`}
                supplier={supplier}
                componentKey={supplier.id}
                allowRemoval={true}
                removeFunction={this.removeSupplier}
              />
            ))
          : 'Unknown'}
      </div>
    );
  }
}
BrandEdit.defaultProps = {
  brand: {},
  suppliers: [],
};
BrandEdit.propTypes = {
  suppliers: PropTypes.array,
  brand: PropTypes.object,
  componentKey: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  deleteBrand: PropTypes.func,
  saveBrand: PropTypes.func,
  closeBrandModal: PropTypes.func,
};
export default BrandEdit;
