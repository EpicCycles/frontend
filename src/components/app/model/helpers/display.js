import {
  BIKE,
  BRAND,
  CHECKBOX,
  CURRENCY,
  CUSTOMER,
  DATE_TIME,
  NUMBER,
  PART_TYPE,
  RADIO,
  SELECT_ONE,
  SUPPLIER,
  USER,
} from './fields';
import { getCustomerName } from '../../../customer/helpers/customer';
import { getBrandName } from '../../../brand/helpers/brand';
import { getNameForValue } from './model';
import { getSupplierName } from '../../../supplier/helpers/supplier';
import { getPartTypeName } from '../../../partType/helpers/partType';
import { getBikeName } from '../../../bike/helpers/bike';
import { getUserName } from '../../../user/helpers/user';
import { CHARGE } from '../../../quoteCharge/helpers/quoteChargeFields';
import { chargeName } from '../../../charge/helpers/chargeName';
import { findObjectWithKey } from '../../../../helpers/utils';
import { fittingText } from '../../../fitting/helpers/fitting';
import { FITTING } from '../../../quote/helpers/quoteFields';

export const fixedHeaderClassname = lockColumn => {
  if (lockColumn) return 'grid-header--fixed-left';
  return '';
};
export const fixedDetailsClassname = lockColumn => {
  if (lockColumn) return 'grid-item--fixed-left';
  return '';
};
export const gridHeaderClass = (baseClassName = '', fieldIndex, firstColumnLocked) => {
  const shouldLock = firstColumnLocked && fieldIndex === 0;
  return `${baseClassName} grid-item--header ${fixedHeaderClassname(shouldLock)}`;
};
export const gridItemClass = (baseClassName = '', fieldIndex, firstColumnLocked) => {
  const shouldLock = firstColumnLocked && fieldIndex === 0;
  return `grid-item ${fixedDetailsClassname(shouldLock)} ${baseClassName}`;
};
export const formattedDate = date => {
  if (date) return date.toLocaleDateString('en-GB');
  return '';
};
export const formattedDateTime = date => {
  if (date) return date.toLocaleString('en-GB');
  return '';
};
export const formattedPrice = fieldValue => {
  if (fieldValue)
    return Number(fieldValue).toLocaleString('en-GB', {
      style: 'currency',
      currency: 'GBP',
    });
  return '';
};
export const fieldAlignment = field => {
  switch (field.type) {
    case NUMBER:
    case CURRENCY:
      return 'align_right';
    case CHECKBOX:
    case RADIO:
      return 'align_center';
    default:
      return '';
  }
};
export const buildViewString = (model, field, sourceDataArrays = {}) => {
  let viewData;
  const {
    sections,
    brands,
    suppliers,
    customers,
    bikes,
    frames,
    users,
    charges,
    fittings,
  } = sourceDataArrays;
  const fieldValue = model ? model[field.fieldName] : undefined;
  switch (field.type) {
    case CUSTOMER:
      viewData = fieldValue ? getCustomerName(fieldValue, customers) : '';
      break;
    case CURRENCY:
      viewData = formattedPrice(fieldValue);
      break;
    case CHECKBOX:
      viewData = fieldValue ? 'Yes' : 'No';
      break;
    case DATE_TIME:
      viewData = fieldValue ? formattedDateTime(new Date(fieldValue)) : '';
      break;
    case PART_TYPE:
      viewData = fieldValue ? getPartTypeName(fieldValue, sections) : '';
      break;
    case BRAND:
      viewData = fieldValue ? getBrandName(fieldValue, brands) : '';
      break;
    case SUPPLIER:
      viewData = fieldValue ? getSupplierName(fieldValue, suppliers) : '';
      break;
    case BIKE:
      viewData = fieldValue ? getBikeName(fieldValue, bikes, frames, brands) : '';
      break;
    case SELECT_ONE:
      viewData = fieldValue ? getNameForValue(fieldValue, field.selectList) : '';
      break;
    case USER:
      viewData = fieldValue ? getUserName(fieldValue, users) : '';
      break;
    case CHARGE:
      viewData = fieldValue ? chargeName(fieldValue, charges) : '';
      break;
    case FITTING:
      const fitting = fieldValue ? findObjectWithKey(fittings, fieldValue) : undefined;
      viewData = fitting ? fittingText(fitting) : '';
      break;
    default:
      viewData = fieldValue ? fieldValue : '';
  }
  return viewData;
};
