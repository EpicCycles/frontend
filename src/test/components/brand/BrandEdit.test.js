import React from 'react';
import toJson from 'enzyme-to-json';
import BrandEdit from '../../../components/brand/BrandEdit';
import { NEW_ELEMENT_ID } from '../../../helpers/constants';
import { BRAND_NAME_MISSING } from '../../../components/app/model/helpers/error';
describe('BrandEdit', () => {
  // props are: brand, componentKey, pickUpBrand
  it('should shows new brand without supplier correctly', () => {
    const brand = { brand_name: 'e brand 8', link: 'https://bianchi.co.uk', id: 8 };
    const componentKey = NEW_ELEMENT_ID;
    const pickUpBrand = jest.fn();
    const component = shallow(
      <BrandEdit brand={brand} componentKey={componentKey} pickUpBrand={pickUpBrand} />,
    );
    expect(toJson(component)).toMatchSnapshot();
  });
  it('should pass a brand change and errors to an existing brand to the supplied function', () => {
    const brand = { brand_name: 'e brand 8', link: 'https://bianchi.co.uk', id: 8, supplier: '34' };
    const componentKey = brand.id;
    const pickUpBrand = jest.fn();
    const handleBrandChange = jest.fn();

    const component = shallow(
      <BrandEdit
        brand={brand}
        componentKey={componentKey}
        pickUpBrand={pickUpBrand}
        handleBrandChange={handleBrandChange}
      />,
    );
    component.instance().handleBrandValueChange('brand_name', '');
    expect(handleBrandChange.mock.calls).toHaveLength(1);
    expect(handleBrandChange.mock.calls[0][0]).toBe(componentKey);
    expect(handleBrandChange.mock.calls[0][1].id).toBe(brand.id);
    expect(handleBrandChange.mock.calls[0][1].brand_name).toBe('');
    expect(handleBrandChange.mock.calls[0][1].error_detail).toEqual({
      brand_name: BRAND_NAME_MISSING,
    });
  });
  it('BrandEdit should set up a new brand as requested', () => {
    const brand = { brand_name: 'e brand 8', link: 'https://bianchi.co.uk', supplier: '34' };
    const componentKey = NEW_ELEMENT_ID;
    const pickUpBrand = jest.fn();
    const handleBrandChange = jest.fn();

    const component = shallow(
      <BrandEdit
        brand={brand}
        componentKey={componentKey}
        pickUpBrand={pickUpBrand}
        handleBrandChange={handleBrandChange}
      />,
    );
    component.instance().addAnother();
    expect(handleBrandChange.mock.calls).toHaveLength(1);
    expect(handleBrandChange.mock.calls[0][0]).toBe(componentKey);
    expect(handleBrandChange.mock.calls[0][1].id).toBe(undefined);
    expect(handleBrandChange.mock.calls[0][1].dummyKey).not.toBe(undefined);
    expect(handleBrandChange.mock.calls[0][1].brand_name).toBe(brand.brand_name);
    expect(handleBrandChange.mock.calls[0][1].link).toBe(brand.link);
  });
  it('should pass a brand change and errors to an new brand to the supplied function', () => {
    const brand = { brand_name: 'e brand 8', link: 'https://bianchi.co.uk', supplier: '34' };
    const componentKey = NEW_ELEMENT_ID;
    const pickUpBrand = jest.fn();
    const handleBrandChange = jest.fn();

    const component = shallow(
      <BrandEdit
        brand={brand}
        componentKey={componentKey}
        pickUpBrand={pickUpBrand}
        handleBrandChange={handleBrandChange}
      />,
    );
    component.instance().handleBrandValueChange('brand_name', '');
    expect(handleBrandChange.mock.calls).toHaveLength(1);
    expect(handleBrandChange.mock.calls[0][0]).toBe(componentKey);
    expect(handleBrandChange.mock.calls[0][1].id).toBe(undefined);
    expect(handleBrandChange.mock.calls[0][1].dummyKey).not.toBe(undefined);
    expect(handleBrandChange.mock.calls[0][1].brand_name).toBe('');
    expect(handleBrandChange.mock.calls[0][1].link).toBe(brand.link);
    expect(handleBrandChange.mock.calls[0][1].error_detail).toEqual({
      brand_name: 'A name is required for the brand',
    });
  });
  it('should pass a brand change to an existing brand to the supplied function', () => {
    const brand = { brand_name: 'e brand 8', link: 'https://bianchi.co.uk', id: 8, supplier: '34' };
    const componentKey = brand.id;
    const pickUpBrand = jest.fn();
    const handleBrandChange = jest.fn();

    const component = shallow(
      <BrandEdit
        brand={brand}
        componentKey={componentKey}
        pickUpBrand={pickUpBrand}
        handleBrandChange={handleBrandChange}
      />,
    );

    component.instance().handleBrandValueChange('brand_name', 'New Name');
    expect(handleBrandChange.mock.calls).toHaveLength(1);
    expect(handleBrandChange.mock.calls[0][0]).toBe(componentKey);
    expect(handleBrandChange.mock.calls[0][1].id).toBe(brand.id);
    expect(handleBrandChange.mock.calls[0][1].brand_name).toBe('New Name');
    expect(handleBrandChange.mock.calls[0][1].error_detail).toEqual({});
  });
  it('should pass a brand change to an new brand to the supplied function', () => {
    const brand = { brand_name: 'e brand 8', link: 'https://bianchi.co.uk', supplier: '34' };
    const componentKey = NEW_ELEMENT_ID;
    const pickUpBrand = jest.fn();
    const handleBrandChange = jest.fn();

    const component = shallow(
      <BrandEdit
        brand={brand}
        componentKey={componentKey}
        pickUpBrand={pickUpBrand}
        handleBrandChange={handleBrandChange}
      />,
    );

    component.instance().handleBrandValueChange('brand_name', 'New Name');
    expect(handleBrandChange.mock.calls).toHaveLength(1);
    expect(handleBrandChange.mock.calls[0][0]).toBe(componentKey);
    expect(handleBrandChange.mock.calls[0][1].id).toBe(undefined);
    expect(handleBrandChange.mock.calls[0][1].dummyKey).not.toBe(undefined);
    expect(handleBrandChange.mock.calls[0][1].brand_name).toBe('New Name');
    expect(handleBrandChange.mock.calls[0][1].link).toBe(brand.link);
    expect(handleBrandChange.mock.calls[0][1].error_detail).toEqual({});
  });
  it('remove supplier when no suppliers does not fail', () => {
    const brand = { brand_name: 'e brand 8', link: 'https://bianchi.co.uk', id: 8 };
    const componentKey = brand.id;
    const pickUpBrand = jest.fn();
    const handleBrandChange = jest.fn();

    const component = shallow(
      <BrandEdit
        brand={brand}
        componentKey={componentKey}
        pickUpBrand={pickUpBrand}
        handleBrandChange={handleBrandChange}
      />,
    );

    component.instance().removeSupplier('25');
    expect(handleBrandChange.mock.calls).toHaveLength(0);
  });
  it('remove supplier when empty array of suppliers does not fail', () => {
    const brand = {
      brand_name: 'e brand 8',
      link: 'https://bianchi.co.uk',
      id: 8,
      supplier: [],
    };
    const componentKey = brand.id;
    const pickUpBrand = jest.fn();
    const handleBrandChange = jest.fn();

    const component = shallow(
      <BrandEdit
        brand={brand}
        componentKey={componentKey}
        pickUpBrand={pickUpBrand}
        handleBrandChange={handleBrandChange}
      />,
    );

    component.instance().removeSupplier('25');
    expect(handleBrandChange.mock.calls).toHaveLength(0);
  });
  it('remove supplier when first in array of suppliers does not fail', () => {
    const brand = {
      brand_name: 'e brand 8',
      link: 'https://bianchi.co.uk',
      id: 8,
      supplier: [23, 1, 2, 45, 16],
    };
    const componentKey = brand.id;
    const pickUpBrand = jest.fn();
    const handleBrandChange = jest.fn();

    const component = shallow(
      <BrandEdit
        brand={brand}
        componentKey={componentKey}
        pickUpBrand={pickUpBrand}
        handleBrandChange={handleBrandChange}
      />,
    );

    component.instance().removeSupplier(23);
    expect(handleBrandChange.mock.calls).toHaveLength(1);
    expect(handleBrandChange.mock.calls[0][0]).toBe(componentKey);
    expect(handleBrandChange.mock.calls[0][1].id).toBe(brand.id);
    expect(handleBrandChange.mock.calls[0][1].dummyKey).toBe(undefined);
    expect(handleBrandChange.mock.calls[0][1].brand_name).toBe(brand.brand_name);
    expect(handleBrandChange.mock.calls[0][1].link).toBe(brand.link);
    expect(handleBrandChange.mock.calls[0][1].supplier).toEqual([1, 2, 45, 16]);
  });
  it('remove supplier when last in array of suppliers does not fail', () => {
    const brand = {
      brand_name: 'e brand 8',
      link: 'https://bianchi.co.uk',
      id: 8,
      supplier: [1, 2, 45, 16, 23],
    };
    const componentKey = brand.id;
    const pickUpBrand = jest.fn();
    const handleBrandChange = jest.fn();

    const component = shallow(
      <BrandEdit
        brand={brand}
        componentKey={componentKey}
        pickUpBrand={pickUpBrand}
        handleBrandChange={handleBrandChange}
      />,
    );

    component.instance().removeSupplier(23);
    expect(handleBrandChange.mock.calls).toHaveLength(1);
    expect(handleBrandChange.mock.calls[0][0]).toBe(componentKey);
    expect(handleBrandChange.mock.calls[0][1].id).toBe(brand.id);
    expect(handleBrandChange.mock.calls[0][1].dummyKey).toBe(undefined);
    expect(handleBrandChange.mock.calls[0][1].brand_name).toBe(brand.brand_name);
    expect(handleBrandChange.mock.calls[0][1].link).toBe(brand.link);
    expect(handleBrandChange.mock.calls[0][1].supplier).toEqual([1, 2, 45, 16]);
  });

  it('remove supplier when only one in array of suppliers does not fail', () => {
    const brand = {
      brand_name: 'e brand 8',
      link: 'https://bianchi.co.uk',
      id: 8,
      supplier: [23],
    };
    const componentKey = brand.id;
    const pickUpBrand = jest.fn();
    const handleBrandChange = jest.fn();

    const component = shallow(
      <BrandEdit
        brand={brand}
        componentKey={componentKey}
        pickUpBrand={pickUpBrand}
        handleBrandChange={handleBrandChange}
      />,
    );

    component.instance().removeSupplier(23);
    expect(handleBrandChange.mock.calls).toHaveLength(1);
    expect(handleBrandChange.mock.calls[0][0]).toBe(componentKey);
    expect(handleBrandChange.mock.calls[0][1].id).toBe(brand.id);
    expect(handleBrandChange.mock.calls[0][1].dummyKey).toBe(undefined);
    expect(handleBrandChange.mock.calls[0][1].brand_name).toBe(brand.brand_name);
    expect(handleBrandChange.mock.calls[0][1].link).toBe(brand.link);
    expect(handleBrandChange.mock.calls[0][1].supplier).toEqual([]);
  });

  it('remove supplier when one of array of suppliers does not fail', () => {
    const brand = {
      brand_name: 'e brand 8',
      link: 'https://bianchi.co.uk',
      id: 8,
      supplier: [1, 2, 23, 45, 16],
    };
    const componentKey = brand.id;
    const pickUpBrand = jest.fn();
    const handleBrandChange = jest.fn();

    const component = shallow(
      <BrandEdit
        brand={brand}
        componentKey={componentKey}
        pickUpBrand={pickUpBrand}
        handleBrandChange={handleBrandChange}
      />,
    );

    component.instance().removeSupplier(23);
    expect(handleBrandChange.mock.calls).toHaveLength(1);
    expect(handleBrandChange.mock.calls[0][0]).toBe(componentKey);
    expect(handleBrandChange.mock.calls[0][1].id).toBe(brand.id);
    expect(handleBrandChange.mock.calls[0][1].dummyKey).toBe(undefined);
    expect(handleBrandChange.mock.calls[0][1].brand_name).toBe(brand.brand_name);
    expect(handleBrandChange.mock.calls[0][1].link).toBe(brand.link);
    expect(handleBrandChange.mock.calls[0][1].supplier).toEqual([1, 2, 45, 16]);
  });
});
