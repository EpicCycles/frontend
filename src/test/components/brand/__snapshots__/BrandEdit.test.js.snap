// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrandEdit should shows new brand without supplier correctly 1`] = `
<div
  className="rounded"
  draggable={false}
  key="brandnew"
  onDragStart={[Function]}
>
  <AddLink
    addFunction={[Function]}
    addObjectName="Brand"
  />
  <EditModelInput
    componentKey="new"
    field={
      Object {
        "displaySize": 20,
        "error": "A name is required for the brand",
        "fieldName": "brand_name",
        "header": "Brand Name",
        "maxLength": 50,
        "required": false,
        "synonyms": Array [],
      }
    }
    index=""
    model={
      Object {
        "brand_name": "e brand 8",
        "id": 8,
        "link": "https://bianchi.co.uk",
      }
    }
    onChange={[Function]}
    persistedModel={Object {}}
  />
  <div
    className="row"
  >
    <div
      className="field-label"
    >
      Bike Brand
    </div>
    <EditModelInput
      componentKey="new"
      field={
        Object {
          "fieldName": "bike_brand",
          "header": "Bike Brand",
          "type": "checkbox",
        }
      }
      index=""
      model={
        Object {
          "brand_name": "e brand 8",
          "id": 8,
          "link": "https://bianchi.co.uk",
        }
      }
      onChange={[Function]}
      persistedModel={Object {}}
    />
  </div>
  Supplier(s):
   
  Unknown
</div>
`;
